cmd/soroban-rpc/internal/ledgerentry_storage/db.go

```diff
@@ -0,0 +1,295 @@
+package ledgerentry_storage
+
+import (
+	"context"
+	"database/sql"
+	"embed"
+	"fmt"
+	"strconv"
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/jmoiron/sqlx"
+	_ "github.com/mattn/go-sqlite3"
+	migrate "github.com/rubenv/sql-migrate"
+	"github.com/stellar/go/support/errors"
+	"github.com/stellar/go/xdr"
+)
+
+//go:embed migrations/*.sql
+var migrations embed.FS
+
+var ErrEmptyDB = errors.New("DB is empty")
+
+const (
+	ledgerEntriesTableName      = "ledger_entries"
+	ledgerEntriesMetaTableName  = "ledger_entries_meta"
+	latestLedgerSequenceMetaKey = "LatestLedgerSequence"
+)
+
+type DB interface {
+	LedgerEntryStorage
+	GetLatestLedgerSequence() (uint32, error)
+	NewLedgerEntryUpdaterTx(forLedgerSequence uint32, maxBatchSize int) (LedgerEntryUpdaterTx, error)
+}
+
+type LedgerEntryUpdaterTx interface {
+	UpsertLedgerEntry(key xdr.LedgerKey, entry xdr.LedgerEntry) error
+	DeleteLedgerEntry(key xdr.LedgerKey) error
+	Done() error
+}
+
+type sqlDB struct {
+	db *sqlx.DB
+}
+
+func OpenSQLiteDB(dbFilePath string) (DB, error) {
+	db, err := sqlx.Open("sqlite3", dbFilePath)
+	if err != nil {
+		return nil, errors.Wrap(err, "open failed")
+	}
+
+	ret := &sqlDB{
+		db: db,
+	}
+
+	err = runMigrations(ret.db.DB, "sqlite3")
+
+	return ret, err
+}
+
+func getLedgerEntry(tx *sqlx.Tx, buffer *xdr.EncodingBuffer, key xdr.LedgerKey) (xdr.LedgerEntry, error) {
+	encodedKey, err := encodeLedgerKey(buffer, key)
+	if err != nil {
+		return xdr.LedgerEntry{}, err
+	}
+
+	sqlStr, args, err := sq.Select("entry").From(ledgerEntriesTableName).Where(sq.Eq{"key": encodedKey}).ToSql()
+	if err != nil {
+		return xdr.LedgerEntry{}, err
+	}
+	var results []string
+	if err := tx.Select(&results, sqlStr, args...); err != nil {
+		return xdr.LedgerEntry{}, err
+	}
+	if len(results) != 1 {
+		return xdr.LedgerEntry{}, sql.ErrNoRows
+	}
+	ledgerEntryBase64 := results[0]
+	var result xdr.LedgerEntry
+	err = xdr.SafeUnmarshalBase64(ledgerEntryBase64, &result)
+	return result, err
+}
+
+func flushLedgerEntryBatch(tx *sqlx.Tx, encodedKeyEntries map[string]*string) error {
```

Comment:
could you modify the encodedKeyEntries to be an array rather then a map ?
while I do like maps, iteration over maps yields unpredictable order. That means that the entries in the database would have different ordering for the same data. ideally, this would not be the case.

Reply:
A map is used to coalesce multiple modifications to the same key (the same key can be modified multiple times or deleted). You cannot do that with a slice.

Reply:
Also, with a slice you cannot batch queries since updates and deletions would need to be done sequentially and in order.

Reply:
sure; we can use a data structure like the following, which would allow us to be persistent:
```golang
type encodedKeyEntry struct {
  key string
  val *string
}
type encodedKeyEntries struct {
   keys []encodedKeyEntry
   keysMaps map[string]uint // translation of key -> index in keys array
}
```

Reply:
as for the batching : 
for update from archives, there are no deletions, so it's very much consistent.
when we write ledger entries, we always want to batch these on ledger boundary ( which would guarantee the same outcome.. )

Reply:
I think this overcomplicates things just to ensure a fixed traversal ordering (which isn’t important for the queries we use)